protocol/superc20/mirror-factory-limited.md

```diff
@@ -0,0 +1,90 @@
+
+The salt doesn't need to be more than the `OptimismMintableERC20Token` address. The `name` and `symbol`, and `decimals` can be fetched with the `token` address alone in the implementation. The factory will be the `Mirror`'s admin, and the proxy will have to either call `setStorage` immediately with the `token` at the time of deployment, or add a `TOKEN_SLOT` and update it on the constructor.
+
+The factory should be upgradable or have setters to allow the modification of the implementations for the current `Mirrors`. The deployment and update of already-deployed `Mirrors` to new implementations should be open to anyone.
+
+We can track versioning through numbers or the addresses of the implementations used.
+
+This design allows the `StandardBridge` or any contract that needs to check the legitimacy of a `Mirror` to call `MIRROR_FACTORY.getMirrorAddress()` and validate whether the calling user is a mirror that corresponds to a given `token`. As an illustrative example of a potential function of the `StandardBridge`:
+
+```solidity
+function mirrorMint(address _token, uint256 _amount) external {
+	address _mirror = MIRROR_FACTORY.getMirrorAddress(_token);
+	if (_mirror != msg.sender) revert();
+	///...mint logic
+}
+
+```
+
+Example code:
+
+```solidity
+contract MirrorFactory {
+	string public constant MIRROR_VERSION = "1.0.0";
+	Mirror public constant MIRROR_IMPL = <PLACEHOLDER_ADDRESS>;
+
+	function deployMirror(address _token) external returns (address) {
+		require(_isOptimismMintableERC20(_token), "not OptimismMintableERC20Token");
+
+		// Note: There's no _token slot in this proxy
+		Proxy _mirror = address(new L2ChugSplashProxy{salt: keccak256(abi.encode(_token))}(address(this), _token));
+
+		// Note: This function does not exist on L2ChugSplashProxy.
+		_mirror.setImplementation(MIRROR_IMPL);
+	}
+
+	function upgradeMirror(address _mirror) external {
+		// Note: This function does not exist on L2ChugSplashProxy.
+		_mirror.setImplementation(MIRROR_IMPL);
+	}
+
+	function _isOptimismMintableERC20(address _token) internal view returns (bool) {
+        return ERC165Checker.supportsInterface(_token, type(ILegacyMintableERC20).interfaceId)
+            || ERC165Checker.supportsInterface(_token, type(IOptimismMintableERC20).interfaceId);
+    }
+
+	function getMirrorAddress(address _token) external view returns (address _mirror) {
+		_mirror = address(uint160(uint(keccak256(abi.encodePacked(
+            bytes1(0xff),
+            address(this),
+            keccak256(abi.encode(_token)),
+            keccak256(abi.encodePacked(
+                type(L2ChugSplashProxy).creationCode,
+                abi.encode(_token)
+            ))
+        )))))
+	} 
+}
+
+```
+
+# Risks & Uncertainties
+
+This only accounts for `OptimismMintableERC20Tokens`. The design space is limited for simplicity. Legacy tokens are not considered. New interoperable tokens using the `Superc20` standard without Mirrors, are not taken into account for this factory.
```

Comment:
This is a user story I'd like to see fleshed out more, as the high level paths are unclear to me:
- Token A has existing OptimismMintableERC20 contract -> describe the path to create a SuperERC20 version
- Token B has no existing OptimismMintableERC20 contract -> describe the path to create a SuperERC20 version
  - Token B now has SuperERC20 version but no OptimismMintableERC20 version -> can someone create the OptimismMintableERC20 now? If yes, what happens? If no, what prevents it?